### PR TITLE
Added support for list of tensors as metric input 

### LINKED ISSUE #2013: Support list of tensors in Metrics
## 🚀 Feature

Related to discussion from here: https://github.com/Project-MONAI/MONAI/issues/2220#issuecomment-849250700

The feature request is to support list of tensors or scalars as input for metrics:
a) classification problem:
```python
acc = Accuracy()
output = (
    [torch.rand(10), torch.rand(10), torch.rand(10)],  # predictions
    [4, 5, 1],  # ground truth
)
acc.update(output)
print(acc.compute())
```

b) segmentation problem:
```python
acc = Accuracy()
output = (
    [torch.rand(10, 32, 32), torch.rand(10, 40, 30), torch.rand(10, 30, 24)],  # predictions
    [torch.randint(0, 10, size=(32, 32)), torch.randint(0, 10, size=(40, 30)),  torch.randint(0, 10, size=(30, 24))],  # ground truth
)
acc.update(output)
print(acc.compute())
```

c) Same for API with Engine.

cc @Nic-Ma

### PR DESCRIPTION
**Pull request by vfdev-5**
Fixes #2013

Description:
- Added support for list of tensors as metric input

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [ ] Documentation is updated (if required)

cc @Nic-Ma for visibility

### CODE REVIEW

**Review by Nic-Ma — APPROVED** [review]

File: ignite/metrics/metric.py [hunk 1]
```diff
@@ -296,7 +305,20 @@ def iteration_completed(self, engine: Engine) -> None:
...
+        if isinstance(output, Sequence) and all([_is_list_of_tensors_or_numbers(o) for o in output]):
+            if not (len(output) == 2 and len(output[0]) == len(output[1])):
+                raise ValueError(
+                    f"Output should have 2 items of the same length, "
+                    f"got {len(output)} and {len(output[0])}, {len(output[1])}"
+                )
+            for o1, o2 in zip(output[0], output[1]):
+                # o1 and o2 are list of tensors or numbers
+                tensor_o1 = _to_batched_tensor(o1)
+                tensor_o2 = _to_batched_tensor(o2, device=tensor_o1.device)
```
Comment (sdesrozis) [inline ignite/metrics/metric.py:318]: Should we move the output to a device or let this be done in the training step function ?

**Comment by sdesrozis** [comment]
Just one remark but it looks good !

File: ignite/metrics/metric.py (hunk 1 above)
Comment (vfdev-5) [reply ignite/metrics/metric.py:318]: here output is something coming from training step and in this case, o2 is a number that have to be wrapped with a tensor and put to appropriate device...

Comment (sdesrozis) [reply ignite/metrics/metric.py:318]: My bad !!

**Review by sdesrozis — APPROVED** [review]
@vfdev-5 Ok !

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 @sdesrozis ,

Thanks for the quick PR, looks very good to me.
We also added support for `list of tensors` to MONAI last week: https://github.com/Project-MONAI/MONAI/blob/dev/monai/metrics/metric.py#L61
I have a quick question about the metrics support for **dict**:
Is it possible to make the global variable `required_output_keys` of ignite Metric to be an arg of `Metric.__init__()`?
We already received much feedback that our workflow programs have too many lambda functions and hard to understand. I am trying to reduce the `output_transform= lambda x: ...` in the code. If we can explicitly specify the `keys` of `y_pred` and `y`, maybe it's easier to use and understand?
CC @wyli

Thanks.

**Comment by vfdev-5** [comment]
Hi @Nic-Ma , 

> Is it possible to make the global variable required_output_keys of ignite Metric to be an arg of Metric.__init__()?

thanks for the feedback, `required_output_keys` is a class attribute and I think it would be better to set `required_output_keys` globally instead of for each metric: 
```python
from ignite.engine import Engine
from ignite.metrics import Accuracy, Metric

Metric.required_output_keys = ("my_y_pred", "my_y_true")

acc = Accuracy()
acc.required_output_keys

engine = Engine(lambda e, b: None)
engine.state.output = {"my_y_pred": torch.rand(4, 10), "my_y_true": torch.randint(0, 10, size=(4, ))}

acc.iteration_completed(engine)
acc.compute()
> 0.125
```
what do you think ?

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

Thanks for your proposal and sample code!
It can solve our problem, but maybe a little bit "hacking", especially when setting ignite args in the **JSON config** of NVIDIA Clara-Train SDK.
Our use case is something like:
```
"trainer": {
    "name": "SupervisedTrainer",
    "args": {
        ... ...
    }
},
post_transforms: [
    {
        "name": "CopyItemsD":
        "args": {"keys": "pred", "times": 1, "names": ["pred_inverted"]}
    },
    {
        "name": "InvertD",
        "args": {"keys": "pred_inverted"}
    },
    {
        "name": "SplitChannelD",
        "args": {"keys": "pred_inverted", "output_postfixes": ["background", "class1", "class2"]}
    }
]
"metrics": [
    {
        "name": "Accuracy",  # ignite metric
        "args": {"output_keys": ["pred", "label"]}
    },
    {
        "name": "MeanDice",  # subclass of ignite metric
        "args": {"output_keys": ["pred_inverted", "label_inverted"]}
    },
    {
        "name": "MeanDice",  # metric on class 1
        "args": {"output_keys": ["pred_inverted_class1", "label_inverted_class1"]}
    },
    {
        "name": "MeanDice",  # metric on class 2
        "args": {"output_keys": ["pred_inverted_class2", "label_inverted_class2"]}
    }
]
```
What do you think is the best way to support it?

Thanks.

**Comment by vfdev-5** [comment]
Thanks for the details, Nic ! That's definitely requires `output_transform` to handle each case. Let me think about that.

> We already received much feedback that our workflow programs have too many lambda functions and hard to understand. I am trying to reduce the output_transform= lambda x: ... in the code.

In this case, maybe, it is possible to introduce one function to deal with those required for metrics like:
```python
def from_dict(*keys):
    def wrapper(output: dict):
        return tuple(output[k] for k in keys)
    return wrapper

acc = Accuracy(output_transform=from_dict("pred_inverted", "label_inverted"))
```
what do you think ?

**Comment by Nic-Ma** [comment]
OK, thanks for the solution.
Let me try to add this API in MONAI utilities.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

BTW, may I know when do you plan to release ignite v0.4.5?

Thanks.

**Comment by vfdev-5** [comment]
Hi Nic,
It should be done by this friday.

**Comment by Nic-Ma** [comment]
Sounds great!
Thanks for letting me know.

**Pull request merged**